tfregistry/provider.go

```diff
@@ -0,0 +1,76 @@
+package tfregistry
+
+import (
+	"context"
+	"fmt"
+)
+
+const (
+	// providerV1Service is a sub path of provider v1 service endpoint.
+	// The service discovery protocol is not implemented for now.
+	// https://www.terraform.io/docs/internals/provider-registry-protocol.html#service-discovery
+	//
+	// Include slashes for later implementation of service discovery.
+	// curl https://registry.terraform.io/.well-known/terraform.json
+	// {"modules.v1":"/v1/modules/","providers.v1":"/v1/providers/"}
+	providerV1Service = "/v1/providers/"
+)
+
+// ProviderV1API is an interface for the provider v1 service.
+type ProviderV1API interface {
+	// ProviderLatest returns the latest version of a provider.
+	// https://www.terraform.io/docs/internals/provider-registry-protocol.html
+	ProviderLatest(ctx context.Context, req *ProviderLatestRequest) (*ProviderLatestResponse, error)
+}
+
+// ProviderLatestRequest is a request parameter for ProviderLatest().
+// https://www.terraform.io/docs/internals/provider-registry-protocol.html#provider-addresses
+type ProviderLatestRequest struct {
+	// Namespace is the name of a namespace, unique on a particular hostname, that can contain one or more providers that are somehow related. On the public Terraform Registry the "namespace" represents the organization that is packaging and distributing the provider.
+	Namespace string `json:"namespace"`
+	// Type is the provider type, like "azurerm", "aws", "google", "dns", etc. A provider type is unique within a particular hostname and namespace.
+	Type string `json:"type"`
+}
+
+// ProviderLatestResponse is a response data for ProviderLatest().
+// There are other response fields, but we define only those we need here.
+type ProviderLatestResponse struct {
+	// Version is the latest version of the provider.
+	Version string `json:"version"`
+	// Versions is a list of available versions.
+	Versions []string `json:"versions"`
+}
+
+// ProviderLatest returns the latest version of a provider.
+// https://www.terraform.io/docs/internals/provider-registry-protocol.html#list-available-versions
```

Comment:
This link is for `:namespace/:type/versions`, not for `:namespace/:type`, and the later seems to be undocumented🤔 
I think it's a reasonable compromise here to use the undocumented API to simplify the implementation.
Could you remove the incorrect link and add a note that we are depending on the undocumented API which may be unstable to simplify the implementation?

Reply:
The same link is copied to some places, so don't forget to fix all.

Reply:
Good point. I just assumed that feature was modeled just like the equivalent latest module feature.
I've opened a topic on the Hashicorp community forum to request more information about that undocumented providers API endpoint: https://discuss.hashicorp.com/t/terraform-registry-api-documentation-for-extended-providers-endpoint-features/14738

In the mean time, what about:
```suggestion
// This relies on a currently undocumented providers API endpoint which behaves exactly like the equivalent documented modules API endpoint.
// https://www.terraform.io/docs/registry/api.html#latest-version-for-a-specific-module-provider
```

Reply:
Thank you for opening the topic on forum.

Even if the undocumented provider latest API is removed in the future, we can calculate the latest version from the documented provider versions API and switch the implementation if we need. However, we probably need a bit complex logic such as sorting semantic versions and excluding non-stable alpha/beta/rc releases. So I don't want to implement it until we actually need it. 

The suggested change for the code comment looks good to me. Could you fix all links?

Reply:
Hi @minamijoyo, [Martin suggested](https://discuss.hashicorp.com/t/terraform-registry-api-documentation-for-extended-providers-endpoint-features/14738/2?u=pdecat) that tfupdate should only use official API endpoints that would be guaranteed to be supported by all registries, not just `registry.terraform.io`. I know that host is currently hardcoded in tfupdate's code but what about allowing to target other hosts by optionally passing it in the provider definition, e.g.:

```
tfupdate release list -s tfregistryProvider registry.terraform.io/hashicorp/aws
```

We could probably reuse / reimplement the version selection logic from the terraform CLI directly: https://github.com/hashicorp/terraform/blob/v0.13.3/internal/getproviders/registry_client.go#L364-L400

WDYT?

Reply:
I didn't aware the definition of the minimum requirement of Registry API. 🤔 
I checked commit history and found that the doc is added by Martin a few months ago before v0.13 release 😉 
https://github.com/hashicorp/terraform/pull/25181
The doc looks a contract between Terraform CLI and Registry as if the Terraform core team says "Don't break the contract for backward compatibility." So we can assume that it's stable.

Although I don't think a third-party registry implementation is widely used and I don't have any plan to support it for now, but it makes sense that the tfupdate can be extended to support all valid provider addresses including third-party registry for future extension.  So I agree to implement using the minimum requirement API.

As implementation, we should not directly imports a specific Terraform version because the tfupdate should support multiple Terraform versions by nature. The implementation of registry protocol in hashicorp/terraform seems to be a bit inconsistent between provider and module, but we can probably use https://github.com/apparentlymart/go-versions for sorting versions which is used by provider protocol implementation.

Reply:
Since switching API requires almost rewriting the current implementation in tfregistry package, If you want to get this feature as soon as possible, I don't mind to merge the current implementation as an experimental feature.

Reply:
> If you want to get this feature as soon as possible, I don't mind a perfect stability for a first step and I can merge the current implementation as an experimental feature.

Ok, let's do this for now.